protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnMapper.java

```diff
@@ -35,54 +37,101 @@
     public final static int FIRST_MSN = 1;
 
     private final HashBiMap<Integer, MessageUid> msnToUid;
+    private final ReadWriteLock readWriteLock;
 
     public UidMsnMapper() {
+        readWriteLock = new ReentrantReadWriteLock();
```

Comment:
The idea is that the underlying structure is not stable (because of out-of-order addition handling)

Then we should allow:
 - mono writer access
 - avoid clients from doing reads during writes as they might see partial mapping.
 - Allow several clients to read from the same data in //, as this is safe.

This is exactly the use case of ReadWrite locks.